src/formElementPlugins/listBuilder/src/ApplicationView.tsx

```diff
@@ -0,0 +1,332 @@
+    resetCurrentResponses(inputFields)
+  )
+
+  const [listItems, setListItems] = useState<ListItem[]>(initialValue?.list ?? [])
+  const [selectedListItem, setSelectedListItem] = useState<number | null>(null)
+  const [open, setOpen] = useState(false)
+  const [inputError, setInputError] = useState(false)
+
+  useEffect(() => {
+    onSave({
+      text: createTextString(listItems),
+      list: listItems,
+    })
+  }, [listItems])
+
+  // This is sent to ApplicationViewWrapper and runs instead of the default responseMutation
+  const innerElementUpdate = async ({ variables: response }: { variables: InputResponseField }) => {
+    const newResponses = [...currentInputResponses]
+    newResponses[response.id as number] = response
+    setCurrentInputResponses(newResponses)
+    if (!anyErrorItems(newResponses, inputFields)) setInputError(false)
+  }
+
+  const updateList = async () => {
+    if (anyErrorItems(currentInputResponses, inputFields)) {
+      setInputError(true)
+      return
+    }
+    // Add item
+    if (selectedListItem === null)
+      setListItems([...listItems, buildListItem(currentInputResponses, inputFields)])
+    else {
+      // Or update existing item
+      const newList = [...listItems]
+      newList[selectedListItem] = buildListItem(currentInputResponses, inputFields)
+      setListItems(newList)
+    }
+    setCurrentInputResponses(resetCurrentResponses(inputFields))
+    setOpen(false)
+    setSelectedListItem(null)
+  }
+
+  const editItem = async (index: number) => {
+    setSelectedListItem(index)
+    setCurrentInputResponses(listItems[index])
+    setOpen(true)
+  }
+
+  const deleteItem = async (index: number) => {
+    setListItems(listItems.filter((_, i) => i !== index))
+    setOpen(false)
+    setCurrentInputResponses(resetCurrentResponses(inputFields))
+  }
+
+  const listDisplayProps: ListLayoutProps = {
+    listItems,
+    displayFormat,
+    editItem,
+    deleteItem,
+    fieldTitles: inputFields.map((e: TemplateElement) => e.title),
+    Markdown,
+  }
+
+  const DisplayComponent =
+    displayType === DisplayType.TABLE ? (
+      <ListTableLayout {...listDisplayProps} />
+    ) : (
+      <ListCardLayout {...listDisplayProps} />
+    )
+
+  return (
+    <>
+      <label>
+        <Markdown text={label} semanticComponent="noParagraph" />
+      </label>
+      <Markdown text={description} />
+      <Modal
+        size="tiny"
+        // closeIcon
+        onClose={() => {
+          setOpen(false)
+          setSelectedListItem(null)
+          setCurrentInputResponses(resetCurrentResponses(inputFields))
+          setInputError(false)
+        }}
+        onOpen={() => setOpen(true)}
+        open={open}
+        trigger={<Button primary content={createModalButtonText} />}
+      >
+        <Segment>
+          <Form>
+            <Markdown text={modalText} />
+            {inputFields.map((field: TemplateElement, index: number) => {
+              const element = {
```

Comment:
probably a good idea to extract this out of jsx, maybe even map inputFields into something that can be iterated over here

Reply:
You mean the "element" fields definitions? Yeah sure, I'd planned to organise this a bit better once I had a clearer idea of which fields were going to be coming from where (and which would just have "dummy" values, etc.)

Reply:
a80f5def